google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,180 @@
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+
+/**
+ * An implementation of a {@link StreamResumptionStrategy} for merged rows. This class tracks the
+ * last complete row seen and upon retry can build a request to resume the stream from where it left
+ * off.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class ReadRowsResumptionStrategy<RowT>
+    implements StreamResumptionStrategy<ReadRowsRequest, RowT> {
+  private final RowAdapter<RowT> rowAdapter;
+  private ByteString lastKey = ByteString.EMPTY;
+  // Number of rows processed excluding Marker row.
+  private long numProcessed;
+
+  public ReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new ReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public void onProgress(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // So it can trim the start of the scan, but does not contribute to the row limit.
+    lastKey = rowAdapter.getKey(response);
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Given a request, this implementation will narrow that request to exclude all row keys and
+   * ranges that would produce rows that come before {@link #lastKey}. Furthermore this
+   * implementation takes care to update the row limit of the request to account for all of the
+   * received rows.
+   */
+  @Override
+  public ReadRowsRequest getResumeRequest(ReadRowsRequest request) {
+    // An empty lastKey means that we have not successfully read the first row,
+    // resume with the original request object.
+    if (lastKey.isEmpty()) {
+      return request;
+    }
+    Builder builder = request.toBuilder();
+
+    if (request.getRowsLimit() > 0) {
+      // NOTE: the edge case of request.getRowsLimit() == numProcessed is handled below
+      Preconditions.checkState(
+          request.getRowsLimit() >= numProcessed,
+          "Detected too many rows for the current row limit during a retry.");
+      builder.setRowsLimit(request.getRowsLimit() - numProcessed);
+    }
+
+    // Reset rows to scan.
+    builder.clearRows();
+
+    RowSet.Builder rowSetBuilder = RowSet.newBuilder();
+
+    // Special case: empty query implies full table scan
+    if (request.getRows().getRowKeysList().isEmpty()
+        && request.getRows().getRowRangesList().isEmpty()) {
+
+      request =
```

Comment:
I would keep the original request object untouched if possible.

Reply:
I assigned it to a local variable instead